libexec/rtld-elf/aarch64/rtld_c18n_asm.S

```diff
@@ -44,51 +49,76 @@ ENTRY(_rtld_longjmp)
 	 * This function MUST preserve the value of x1 and update c0 to point at
 	 * the general purpose registers to be restored.
 	 */
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	mrs	c18, rcsp_el0
+#define	TRUSTED_STACK	c18
+#else
 #define	TRUSTED_STACK	csp
+#endif
 
 	mov	c2, TRUSTED_STACK
 
 	str	c30, [TRUSTED_STACK, #-CAP_WIDTH]!
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	msr	rcsp_el0, TRUSTED_STACK
+#endif
 
 	bl	_rtld_longjmp_impl
 
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	mrs	TRUSTED_STACK, rcsp_el0
+#endif
 	ldr	c30, [TRUSTED_STACK], #CAP_WIDTH
 
 	ldr	x3, [TRUSTED_STACK]
 	scvalue	TRUSTED_STACK, TRUSTED_STACK, x3
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	msr	rcsp_el0, TRUSTED_STACK
+#endif
 
 	RETURN
 #undef	TRUSTED_STACK
 END(_rtld_longjmp)
 
 ENTRY(_rtld_thread_start)
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	mov	c10, csp
+	msr	rcsp_el0, c10
+#else
 	/*
 	 * Move the Executive mode thread pointer to Restricted mode.
 	 */
 	mrs	c10, ctpidr_el0
 	msr	rctpidr_el0, c10
+#endif
 	b	_rtld_thread_start_impl
 END(_rtld_thread_start)
 
 ENTRY(_rtld_thr_exit)
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	mrs	c10, rcsp_el0
+	mov	csp, c10
+#endif
 	b	_rtld_thr_exit_impl
 END(_rtld_thr_exit)
 
 ENTRY(_rtld_sighandler)
+#ifndef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
 	mov	c3, csp
+#endif
 	b	_rtld_sighandler_impl
 END(_rtld_sighandler)
 
 ENTRY(_rtld_get_rstk)
 	/*
 	 * This function MUST preserve all caller-saved registers except:
-	 * - c11: Target restricted stack
-	 * - c18: Will be overwritten later in the trampoline
+	 * - c17: Target restricted stack
```

Comment:
This isn't hugely useful, it's the what not the why. If you want to document all the non-standard calling conventions going on, you need to say what registers are being used in non-standard ways on entry and exit, and where they're coming from / being passed onto. Otherwise anyone reading this code has no idea how it fits together and has to trawl through the assembly to figure out the flow of these random registers.